File: 374-GuessNumberHigherOrLower/374-GuessNumberHigherOrLower.js
// Last updated: 4/29/2026, 10:30:43 AM
/** 
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * var guess = function(num) {}
 */

/**
 * @param {number} n
 * @return {number}
 */
var guessNumber = function(n) {
    const dfs = (left, right) => {
        // if(left === right) {
        //     return left;
        // }
        console.log(`left: ${left}, right=${right}`)
        const mid = Math.floor((left + right) / 2);
        const res = guess(mid);
        if(res === 0) {
            return mid;
        } else if(res > 0) {
            return dfs(mid + 1, right);
        } else {
            return dfs(left, mid - 1);
        }
    }

    return dfs(1, n);
};

File: 374-GuessNumberHigherOrLower/374-GuessNumberHigherOrLower.py
# Last updated: 4/29/2026, 10:30:48 AM
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if num is higher than the picked number
#          1 if num is lower than the picked number
#          otherwise return 0
# def guess(num: int) -> int:

class Solution:
    def guessNumber(self, n: int) -> int:
        left = 1
        right = n
        while left <= right:
            mid = (left + right) // 2
            print(f"left: {left}, mid: {mid}, right: {right}")
            if guess(mid) == 0:
                return mid
            if guess(mid) < 0:
                right = mid - 1
            else:
                left = mid + 1
        print(f"left: {left}, right: {right}")
        return -1
        